crates/trans_token/src/tx.rs

```diff
@@ -0,0 +1,548 @@
+    ) -> Option<&Amount> {
+        if key_token == &self.token && key_owner == &self.src_or_dest {
+            return Some(&self.amount);
+        }
+        None
+    }
+
+    fn into_iter(
+        self,
+    ) -> impl IntoIterator<Item = ((Address, Address), Amount)> {
+        [((self.src_or_dest.clone(), self.token.clone()), self.amount)]
+    }
+}
+
+/// Transfer transparent token, insert the verifier expected by the VP and an
+/// emit an event.
+pub fn transfer<ENV>(
+    env: &mut ENV,
+    source: &Address,
+    target: &Address,
+    token: &Address,
+    amount: Amount,
+    event_desc: Cow<'static, str>,
+) -> Result<()>
+where
+    ENV: TxEnv + EmitEvents,
+{
+    multi_transfer(
+        env,
+        SingleCreditOrDebit {
+            src_or_dest: source.clone(),
+            token: token.clone(),
+            amount,
+        },
+        SingleCreditOrDebit {
+            src_or_dest: target.clone(),
+            token: token.clone(),
+            amount,
+        },
+        event_desc,
+    )?;
+
+    Ok(())
+}
+
+#[cfg(test)]
+mod test {
+    use std::collections::BTreeMap;
+
+    use namada_core::address::testing::{
+        arb_address, arb_non_internal_address,
+    };
+    use namada_core::token::testing::arb_amount;
+    use namada_core::uint::Uint;
+    use namada_core::{address, token};
+    use namada_events::extend::EventAttributeEntry;
+    use namada_tests::tx::{ctx, tx_host_env};
+    use proptest::prelude::*;
+
+    use super::*;
+    use crate::event::{PostBalances, SourceAccounts, TargetAccounts};
+
+    const EVENT_DESC: Cow<'static, str> = Cow::Borrowed("event-desc");
+
+    proptest! {
+        #[test]
+        fn test_valid_transfer_tx(
+            (src, dest, token) in (
+                arb_non_internal_address(),
+                arb_non_internal_address(),
+                arb_address()
+            ).prop_filter("unique addresses", |(src, dest, token)|
+                src != dest && dest != token && src != token),
+
+            amount in arb_amount(),
+        ) {
+            test_valid_transfer_tx_aux(src, dest, token, amount)
+        }
+    }
+
+    fn test_valid_transfer_tx_aux(
+        src: Address,
+        dest: Address,
+        token: Address,
+        amount: Amount,
+    ) {
+        tx_host_env::init();
+
+        tx_host_env::with(|tx_env| {
+            tx_env.spawn_accounts([&src, &dest, &token]);
+            tx_env.credit_tokens(&src, &token, amount);
+        });
+        assert_eq!(read_balance(ctx(), &token, &src).unwrap(), amount);
+
+        transfer(ctx(), &src, &dest, &token, amount, EVENT_DESC).unwrap();
+
+        // Dest received the amount
+        assert_eq!(read_balance(ctx(), &token, &dest).unwrap(), amount);
+
+        // Src spent the amount
+        assert_eq!(
+            read_balance(ctx(), &token, &src).unwrap(),
+            token::Amount::zero()
+        );
+
+        // Internal token address have to be part of the verifier set
```

Comment:
Can/should we also check that the src and dest addresses are part of the validation set?

Reply:
Edit, I see that we do that check in the test in crates/token/src/tx.rs, so I'm not sure my comment is still relevant

Reply:
I've added this to be complete and a bit more